### PR TITLE
Bug 559860 - Upgrade to Maven indexer 6.0.0

### PR DESCRIPTION
**Pull request by mbooth101**
Signed-off-by: Mat Booth <mat.booth@redhat.com>

### CODE REVIEW

**Comment by mickaelistria** [comment]
This also requires a change in https://github.com/tesla/m2e-core-tests/ which does use some of the indexer API. Can you please submit a change there too?

**Comment by mbooth101** [comment]
> This also requires a change in https://github.com/tesla/m2e-core-tests/ which does use some of the indexer API. Can you please submit a change there too?

Sure, I always forget about this repo

**Comment by mbooth101** [comment]
@mickaelistria Only a trivial change needed in the test repo: https://github.com/tesla/m2e-core-tests/pull/123

**Comment by mickaelistria** [comment]
Ran https://github.com/eclipse/dash-licenses#usage on `org.eclipse.maven.runtime.indexer` after that change
and got
```
$ mvn dependency:list | grep -Poh "\S+:(system|provided|compile)" | sort | uniq  | java -jar ~/git/dash-licenses/target/org.eclipse.dash.licenses-0.0.1-SNAPSHOT.jar -
Querying Eclipse Foundation for license data for 93 items.
Found 47 items.
Querying ClearlyDefined for license data for 46 items.
Rejected: https://clearlydefined.io/definitions/maven/mavencentral/org.apache.lucene/lucene-sandbox/5.5.5
Found 45 items.
License information could not automatically verified for the following content:

maven/mavencentral/org.codehaus.mojo/animal-sniffer-annotations/1.14 ()
maven/mavencentral/org.apache.lucene/lucene-sandbox/5.5.5 (null)

Please create contribution questionnaires for this content.
```

animal-sniffer 1.14 is already approved with http://dev.eclipse.org/ipzilla/show_bug.cgi?id=19603
I've opening a https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22836 for lucene-sandbox 5.5.5. I'm strongly confident it gets approved, so I'll be merging it after a local test execution (as CI cannot successfully do it because of some tests being in a separate repo...)

**Comment by mickaelistria** [comment]
It looks like that causes 1 test to fail
```
org.codehaus.plexus.component.repository.exception.ComponentLookupException: java.util.NoSuchElementException
      role: org.apache.maven.archetype.source.ArchetypeDataSource
  roleHint: nexus
	at org.eclipse.m2e.tests.archetype.ArchetypeManagerTest.testArchetypeManager(ArchetypeManagerTest.java:71)
Caused by: org.codehaus.plexus.component.repository.exception.ComponentLookupException: 
java.util.NoSuchElementException
      role: org.apache.maven.archetype.source.ArchetypeDataSource
  roleHint: nexus
	at org.eclipse.m2e.tests.archetype.ArchetypeManagerTest.testArchetypeManager(ArchetypeManagerTest.java:71)
Caused by: java.util.NoSuchElementException
	at org.eclipse.m2e.tests.archetype.ArchetypeManagerTest.testArchetypeManager(ArchetypeManagerTest.java:71)
```

To reproduce:
```
$ cd m2e-maven-runtime
$ mvn clean install
$ cd ..    #back to m2e-core root
$ mvn clean install -DskipTests
$ cd m2e-core-tests/org.eclipse.m2e.tests
$ mvn clean verify -Dtest=ArchetypeManaagerTest
```

**Comment by mbooth101** [comment]
@mickaelistria Ah thanks, I believe it should be fixed in this newest revision of the change

File: m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/java/org/apache/maven/index/NexusArchetypeDataSource.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.maven.archetype.source.ArchetypeDataSource;
+import org.apache.maven.index.archetype.AbstractArchetypeDataSource;
+import org.apache.maven.index.context.IndexingContext;
+
+/**
+ * Trivial implementation of a Nexus indexer-based data source for archetypes.
+ */
+@Singleton
+@Named("nexus")
+public class NexusArchetypeDataSource extends AbstractArchetypeDataSource implements ArchetypeDataSource {
```
Comment (mickaelistria) [inline m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/java/org/apache/maven/index/NexusArchetypeDataSource.java:29]: Can you please add a not in the class and/or in the commit message about why this class is now necessary?

File: m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/resources/META-INF/sisu/javax.inject.Named [hunk 2]
```diff
@@ -0,0 +1,20 @@
+org.apache.maven.index.creator.MavenArchetypeArtifactInfoIndexCreator
```
Comment (mickaelistria) [inline m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/resources/META-INF/sisu/javax.inject.Named:1]: Can you please add a not in the class and/or in the commit message about why this class is now necessary?
Couldn't it be taken from somewhere else?

**Review by mickaelistria — CHANGES_REQUESTED** [review]
That seems to work, great!
However, I fail at understanding some of the changes. Can you please comment a bit about them in the files and/or commit message?

File: m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/java/org/apache/maven/index/NexusArchetypeDataSource.java (hunk 1 above)
Comment (mbooth101) [reply m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/java/org/apache/maven/index/NexusArchetypeDataSource.java:35]: I added some comments to explain why this class exists

File: m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/resources/META-INF/sisu/javax.inject.Named (hunk 2 above)
Comment (mbooth101) [reply m2e-maven-runtime/org.eclipse.m2e.maven.indexer/src/main/resources/META-INF/sisu/javax.inject.Named]: Actually I discovered I could adequately generate this file so I added some comments about it to the pom.xml

File: m2e-maven-runtime/org.eclipse.m2e.maven.indexer/pom.xml [hunk 3]
```diff
@@ -50,10 +54,34 @@
...
+      <scope>provided</scope>
+    </dependency>
   </dependencies>
 
   <build>
     <plugins>
+    <!-- Need to regenerate the "javax.inject.Named" index to include additional 
+        injectable classes that do not override those provided by maven-indexer, 
+        this is included in the classpath in front of maven-indexer so it takes precedence 
+        over the one in the maven-indexer jar -->
```
Comment (mbooth101) [inline m2e-maven-runtime/org.eclipse.m2e.maven.indexer/pom.xml:70]: "javax.inject.Named" is now generated, but here is the reason why it exists at all

**Comment by mickaelistria** [comment]
Thanks a lot @mbooth101 , this is now merged and will be available in snapshots shortly.
Note that we won't be able to release until https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22836 is resolved (hopefully very soon).

**Comment by mickaelistria** [comment]
Merged as b443b38924aa529993cb7f0c2481ebdbce157e2f

**Pull request closed without merging**